parsely_raw_data/__init__.py

```diff
@@ -14,17 +14,58 @@
 limitations under the License.
 """
 
-__version__ = '2.3.0.dev0'
+__version__ = "2.3.0"
 
 from . import bigquery, docgen, redshift, s3, samples, schema, stream, utils
+from six import iteritems
 
 __all__ = [
-    'bigquery',
-    'docgen',
-    'redshift',
-    's3',
-    'samples',
-    'schema',
-    'stream',
-    'utils',
+    "bigquery",
+    "docgen",
+    "redshift",
+    "s3",
+    "samples",
+    "schema",
+    "stream",
+    "utils",
 ]
+
+BOOLEAN_FIELDS = ["flags_is_amp"]
+
+
+def normalize_keys(input_event_dict, schema=None):
+    """Conform events to public schema: correct keys and proper value types.
+
+    @param input_event_dict: A dictionary containing Parse.ly pixel events
+    @param schema:  Optional parameter containing the schema to normalize the event_dict keys against
+                    IF not specified, this will default to the most recent parsely_raw_data schema
+    """
+    event_dict = {}
+    schema = schema or schema.SCHEMA
+
+    # fix value types
+    if input_event_dict.get("metadata.share_urls") is not None and isinstance(
+        input_event_dict["metadata.share_urls"], dict
+    ):
+        input_event_dict["metadata.share_urls"] = (
+            list(input_event_dict["metadata.share_urls"].values()) or None
+        )
+
+    # replace all "."s in the key with "_"
+    input_event_dict = {x.replace(',', '_'): v for x, v in input_event_dict.items()}
+
+    # emit only public schema items
+    # ensure all columns are available and null when needed
+    # account for all boolean schema defined fields as this is parsely_raw_data specific
+    for key in schema:
+        if key not in input_event_dict.keys():
```

Comment:
does that check the dictionary values too? if so, i would be hesitant to use it with the amount of metadata free-form text we get

Reply:
That only checks for the presence of the key, not the values themselves. And it's going to operate in constant time because python dicts are stupid fast these days.